src/virtual_machine.rs

```diff
@@ -225,30 +306,71 @@ impl VM {
 
     fn execute(&mut self, operation: Opcode) {
         match operation {
-            Opcode::OP_ADD {
+            Opcode::OpAdd {
                 dr,
                 sr1,
                 second_arg,
             } => {
                 let sr1_val = self.value_from_register(sr1);
                 let second_value = match second_arg {
-                    AddMode::IMMEDIATE { imm5 } => {
+                    Mode::IMMEDIATE { imm5 } => {
                         let value = sign_extend(imm5, 5);
```

Comment:
You´re calling the `sign_extend` function in decode and execute. It's not necessary to call it here.

Reply:
Thanks Dami! Great catch